From: Ben Hutchings <ben@decadent.org.uk>
Subject: epoll: Avoid ABI change in struct file
Date: Sun, 23 Dec 2012 19:05:04 +0000

commit 28d82dc1c4ed ('epoll: limit paths') added the member
f_tfile_llink to the middle of struct file and was backported in Linux
2.6.32.60.  This structure is always allocated in filp_open(), with a
few exceptions that are not exposed to userland (so not used with
epoll), so we can add this member at the end.  Move it there and hide
it from genksyms.

--- a/include/linux/fs.h
+++ b/include/linux/fs.h
@@ -940,12 +940,16 @@ struct file {
 #ifdef CONFIG_EPOLL
 	/* Used by fs/eventpoll.c to link all the hooks to this file */
 	struct list_head	f_ep_links;
-	struct list_head	f_tfile_llink;
 #endif /* #ifdef CONFIG_EPOLL */
 	struct address_space	*f_mapping;
 #ifdef CONFIG_DEBUG_WRITECOUNT
 	unsigned long f_mnt_write_state;
 #endif
+#ifndef __GENKSYMS__
+#ifdef CONFIG_EPOLL
+	struct list_head	f_tfile_llink;
+#endif
+#endif
 };
 extern spinlock_t files_lock;
 #define file_list_lock() spin_lock(&files_lock);
